### PR TITLE
creates run loop for data repair checker

### PR DESCRIPTION
**Pull request by jenlij**

### CODE REVIEW

File: pkg/datarepair/checker/checker.go [hunk 1]
```diff
@@ -120,12 +122,39 @@ func lookupResponsesToNodes(responses *pb.LookupResponses) []*pb.Node {
...
+					}
+				}()
+				return
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	//TODO collectErrors and wait?
+	return utils.CombineErrors(errs...)
```
Comment (coyle) [inline pkg/datarepair/checker/checker.go]: I think this will just return immediately, executing your defers

File: pkg/datarepair/checker/checker.go [hunk 2]
```diff
@@ -120,12 +122,39 @@ func lookupResponsesToNodes(responses *pb.LookupResponses) []*pb.Node {
...
+			select {
+			case <-ticker.C:
+				zap.S().Info("Starting segment checker service")
+				go func() {
+					err := c.IdentifyInjuredSegments(ctx)
+					if err != nil {
+						errs = append(errs, err)
+					}
+				}()
+				return
```
Comment (coyle) [inline pkg/datarepair/checker/checker.go]: this will stop the loop after the first tick, is that what you want?

File: pkg/datarepair/checker/checker.go [hunk 3]
```diff
@@ -120,12 +122,39 @@ func lookupResponsesToNodes(responses *pb.LookupResponses) []*pb.Node {
...
+// Run the checker loop
+func (c *checker) Run(ctx context.Context, interval time.Duration) error {
+	zap.S().Info("Checker is starting up")
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	var errs []error
+
+	go func() {
```
Comment (coyle) [inline pkg/datarepair/checker/checker.go]: nit -- it might be beneficial to make this a named function, outside `Run` so you can test it independently

File: pkg/datarepair/checker/checker.go [hunk 4]
```diff
@@ -120,12 +122,39 @@ func lookupResponsesToNodes(responses *pb.LookupResponses) []*pb.Node {
...
 
-// Run
-func (c *checker) Run() error {
-	return nil
+// Run the checker loop
+func (c *checker) Run(ctx context.Context, interval time.Duration) error {
+	zap.S().Info("Checker is starting up")
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+	ctx, cancel := context.WithCancel(ctx)
```
Comment (coyle) [inline pkg/datarepair/checker/checker.go]: Why do we need the cancel ?

File: pkg/datarepair/checker/checker.go [hunk 5]
```diff
@@ -120,12 +122,39 @@ func lookupResponsesToNodes(responses *pb.LookupResponses) []*pb.Node {
...
+	//TODO collectErrors and wait?
+	return utils.CombineErrors(errs...)
 }
 
-// Stop
-func (c *checker) Stop() error {
+// Stop the checker loop
+func (c *checker) Stop(ctx context.Context) error {
+	_, cancel := context.WithCancel(ctx)
+	cancel()
```
Comment (coyle) [inline pkg/datarepair/checker/checker.go]: this won't do anything. this creates a child context. The cancel only cancels the child not the parent. 

You can add a cancel func to the checker struct and then call `c.cancel()`

File: pkg/datarepair/checker/config.go [hunk 6]
```diff
@@ -7,7 +7,13 @@ import (
 	"context"
 	"time"
 
+	"go.uber.org/zap"
+
+	q "storj.io/storj/pkg/datarepair/queue"
```
Comment (coyle) [inline pkg/datarepair/checker/config.go]: why rename ?

File: pkg/datarepair/checker/config.go [hunk 7]
```diff
@@ -18,7 +24,15 @@ type Config struct {
...
-	return &checker{}, nil
+	pointerdb := pointerdb.LoadFromContext(ctx)
+	overlay := overlay.LoadServerFromContext(ctx)
+	client, err := redis.NewClientFrom(c.QueueAddress)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+	repairQueue := q.NewQueue(client)
+	check := newChecker(pointerdb, repairQueue, overlay, 0, zap.L())
+	return check, nil
```
Comment (coyle) [inline pkg/datarepair/checker/config.go]: nit -- `return newChecker(pointerdb, repairQueue, overlay, 0, zap.L()), nil`

File: pkg/overlay/config.go [hunk 8]
```diff
@@ -36,7 +36,8 @@ type Config struct {
 type CtxKey int
 
 const (
-	ctxKeyOverlay CtxKey = iota
+	ctxKeyOverlay       CtxKey = iota
+	ctxKeyOverlayServer CtxKey = iota + 1
```
Comment (coyle) [inline pkg/overlay/config.go]: https://programming.guide/go/iota.html

File: pkg/overlay/config.go [hunk 9]
```diff
@@ -102,16 +103,18 @@ func (c Config) Run(ctx context.Context, server *provider.Provider) (
...
 
 		// TODO(jt): do something else
 		logger:  zap.L(),
 		metrics: monkit.Default,
-	})
-
-	return server.Run(context.WithValue(ctx, ctxKeyOverlay, cache))
+	}
+	pb.RegisterOverlayServer(server.GRPC(), srv)
+	context.WithValue(ctx, ctxKeyOverlay, cache)
```
Comment (coyle) [inline pkg/overlay/config.go]: this returns a new context with the values added.

https://golang.org/pkg/context/#WithValue

File: pkg/datarepair/checker/checker.go (hunk 1 above)
Comment (dylanlott) [reply pkg/datarepair/checker/checker.go]: I'm moving CollectErrors into pkg/utils. I think it would be a good idea to use that here.

File: pkg/datarepair/checker/checker.go (hunk 3 above)
Comment (egonelbre) [reply pkg/datarepair/checker/checker.go]: Or remove the goroutine completely?

File: pkg/datarepair/checker/checker.go (hunk 2 above)
Comment (jenlij) [reply pkg/datarepair/checker/checker.go]: nope I don't want that. rewriting.

File: pkg/datarepair/checker/checker.go (hunk 5 above)
Comment (jenlij) [reply pkg/datarepair/checker/checker.go]: gotcha I was concerned that we didn't want context on the checker struct

File: pkg/datarepair/checker/checker.go [hunk 10]
```diff
@@ -14,23 +15,27 @@ import (
...
 // Checker is the interface for the data repair queue
 type Checker interface {
 	IdentifyInjuredSegments(ctx context.Context) (err error)
 	Run() error
-	Stop() error
 }
 
 // Checker contains the information needed to do checks for missing pieces
 type checker struct {
+	ctx         context.Context
```
Comment (coyle) [inline pkg/datarepair/checker/checker.go]: I don't think it's typically a good idea to put the context on the struct. We should pass it between functions.

File: pkg/datarepair/checker/checker.go [hunk 11]
```diff
@@ -120,12 +125,29 @@ func lookupResponsesToNodes(responses *pb.LookupResponses) []*pb.Node {
...
-// Stop
-func (c *checker) Stop() error {
-	return nil
+	go func() {
+		for range ticker.C {
+			zap.S().Info("Starting segment checker service")
+			go func() {
+				err := c.IdentifyInjuredSegments(c.ctx)
+				if err != nil {
+					c.errs = append(c.errs, err)
```
Comment (coyle) [inline pkg/datarepair/checker/checker.go]: This is possibly racy.

https://medium.com/@cep21/gos-append-is-not-always-thread-safe-a3034db7975

File: pkg/datarepair/checker/config.go [hunk 12]
```diff
@@ -18,7 +24,18 @@ type Config struct {
...
 func (c Config) initialize(ctx context.Context) (Checker, error) {
-	return &checker{}, nil
+	pointerdb := pointerdb.LoadFromContext(ctx)
+	overlay := overlay.LoadServerFromContext(ctx)
+	client, err := redis.NewClientFrom(c.QueueAddress)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+	repairQueue := queue.NewQueue(client)
+	check := newChecker(pointerdb, repairQueue, overlay, 0, zap.L())
```
Comment (coyle) [inline pkg/datarepair/checker/config.go]: I think we shouldn't have to do this setup after calling new checker. Could we not do this in new checker itself ?

File: pkg/overlay/config.go [hunk 13]
```diff
@@ -102,16 +103,18 @@ func (c Config) Run(ctx context.Context, server *provider.Provider) (
...
+	context.WithValue(ctx, ctxKeyOverlayServer, srv)
+	return server.Run(ctx)
```
Comment (coyle) [inline pkg/overlay/config.go]: this context won't have the values you've added.

File: pkg/datarepair/checker/checker.go (hunk 10 above)
Comment (aleitner) [reply pkg/datarepair/checker/checker.go]: Yeah. We are changing the repairer to just pass the ctx to the repairier.Run() method. Checker should also do that

File: pkg/datarepair/repairer/repairer.go [hunk 14]
```diff
@@ -45,23 +45,22 @@ func (r *repairer) Run(ctx context.Context) (err error) {
...
+				err := r.Repair(ctx, &seg)
+				if err != nil {
+					zap.L().Error("Repair failed", zap.Error(err))
+				}
+			})
 		case <-ctx.Done(): // or the repairer is canceled via context
 			return ctx.Err()
 		}
-
-		seg, err := r.queue.Dequeue()
```
Comment (egonelbre) [inline pkg/datarepair/repairer/repairer.go]: Why was this part changed? The reason it was this way separated is to make clear what select is waiting on.
